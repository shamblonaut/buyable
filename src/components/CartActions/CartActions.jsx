import { Trash2, ShoppingCart } from "lucide-react";
import PropTypes from "prop-types";

import { ProductType } from "@/utils/types";

import { QuantitySelector } from "@/components";
import { Container, AddButton, RemoveButton } from "./CartActions.styles";

const CartActions = ({ product, cart, setCart }) => {
  const addToCart = () => {
    setCart((cart) => {
      const newCart = { ...cart };
      newCart[product.id] = 1;
      return newCart;
    });
  };

  const removeFromCart = () => {
    setCart((cart) => {
      const newCart = { ...cart };
      delete newCart[product.id];
      return newCart;
    });
  };

  const incrementCartQuantity = () => {
    setCart((cart) => {
      const newCart = { ...cart };
      newCart[product.id] += 1;
      return newCart;
    });
  };

  const decrementCartQuantity = () => {
    setCart((cart) => {
      const newCart = { ...cart };
      newCart[product.id] -= 1;

      if (newCart[product.id] === 0) {
        delete newCart[product.id];
      }

      return newCart;
    });
  };

  const setQuantity = (newQuantity) => {
    if (newQuantity <= 0) {
      removeFromCart();
      return;
    }

    setCart((cart) => {
      const newCart = { ...cart };
      newCart[product.id] = newQuantity;
      return newCart;
    });
  };

  return (
    <Container>
      {Object.keys(cart).includes(product.id.toString()) ? (
        <>
          <QuantitySelector
            quantity={cart[product.id]}
            setQuantity={setQuantity}
            increment={incrementCartQuantity}
            decrement={decrementCartQuantity}
          />
          <RemoveButton onClick={removeFromCart}>
            <Trash2 />
          </RemoveButton>
        </>
      ) : (
        <AddButton onClick={addToCart}>
          <ShoppingCart /> Add to cart
        </AddButton>
      )}
    </Container>
  );
};

CartActions.propTypes = {
  product: ProductType.isRequired,
  cart: PropTypes.object.isRequired,
  setCart: PropTypes.func.isRequired,
};

export default CartActions;
