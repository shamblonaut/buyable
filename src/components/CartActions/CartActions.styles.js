import styled from "styled-components";

export const Container = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;

  padding: 8px;
  width: 75%;
`;

const Button = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 8px;

  color: var(--color-light);
  border: 2px solid var(--color-dark);
  border-radius: 4px;

  transition: all 200ms ease;
`;

export const AddButton = styled(Button)`
  background-color: var(--color-green);
  width: 100%;
  padding: 8px 32px;

  &:hover {
    background-color: var(--color-green-dark);
    color: var(--color-light);
  }

  & .lucide {
    margin-right: 8px;
  }
`;

export const RemoveButton = styled(Button)`
  margin-left: 8px;
  background-color: var(--color-red);

  &:hover {
    background-color: var(--color-red-dark);
    color: var(--color-light);
  }
`;
