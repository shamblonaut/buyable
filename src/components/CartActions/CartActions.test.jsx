import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import CartActions from "./CartActions";

const mockCart = { 7: 13 };
const mockProduct = { id: 7 };

const mockSetCart = vi.fn();

describe("CartActions component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders correctly when cart is empty", () => {
    const { container } = render(
      <CartActions product={mockProduct} cart={{}} setCart={mockSetCart} />,
    );

    expect(container).toMatchSnapshot();
  });

  it("renders correctly when cart is not empty", () => {
    const { container } = render(
      <CartActions
        product={mockProduct}
        cart={mockCart}
        setCart={mockSetCart}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("shows add to cart button when cart is empty", () => {
    render(
      <CartActions product={mockProduct} cart={{}} setCart={mockSetCart} />,
    );

    expect(
      within(screen.getByRole("button", { name: "cart-add-button" })).getByText(
        "add to cart",
        { exact: false },
      ),
    ).toBeInTheDocument();
    expect(
      screen.queryByRole("button", { name: "cart-remove-button" }),
    ).not.toBeInTheDocument();
  });

  it("shows remove from cart button and quantity selector when cart is not empty", () => {
    render(
      <CartActions
        product={mockProduct}
        cart={mockCart}
        setCart={mockSetCart}
      />,
    );

    expect(
      screen.getByRole("button", { name: "cart-remove-button" }),
    ).toBeInTheDocument();
    expect(
      screen.queryByRole("button", { name: "cart-add-button" }),
    ).not.toBeInTheDocument();
    expect(screen.getByLabelText("cart-quantity-selector")).toBeInTheDocument();
  });

  it("shows no. of items of current product in quantity selector", () => {
    render(
      <CartActions
        product={mockProduct}
        cart={mockCart}
        setCart={mockSetCart}
      />,
    );

    expect(
      within(screen.getByLabelText("cart-quantity-selector")).getByRole(
        "textbox",
      ),
    ).toHaveValue(mockCart[mockProduct.id].toString());
  });

  it("adds an item to the cart when add to cart is clicked", async () => {
    const user = userEvent.setup();

    let cart = {};
    let setCart = vi.spyOn(
      {
        setCart: (newCart) => {
          if (typeof newCart === "function") {
            cart = newCart(cart);
          } else {
            cart = newCart;
          }
        },
      },
      "setCart",
    );

    const { rerender } = render(
      <CartActions product={mockProduct} cart={cart} setCart={setCart} />,
    );

    await user.click(screen.getByRole("button", { name: "cart-add-button" }));

    expect(setCart).toHaveBeenCalled();
    expect(cart).toMatchObject({ 7: 1 });

    rerender(
      <CartActions product={mockProduct} cart={cart} setCart={setCart} />,
    );

    expect(
      within(screen.getByLabelText("cart-quantity-selector")).getByRole(
        "textbox",
      ),
    ).toHaveValue("1");
  });

  it("removes item from the cart when remove from cart is clicked", async () => {
    const user = userEvent.setup();

    let cart = { ...mockCart };
    let setCart = vi.spyOn(
      {
        setCart: (newCart) => {
          if (typeof newCart === "function") {
            cart = newCart(cart);
          } else {
            cart = newCart;
          }
        },
      },
      "setCart",
    );

    const { rerender } = render(
      <CartActions product={mockProduct} cart={cart} setCart={setCart} />,
    );

    await user.click(
      screen.getByRole("button", { name: "cart-remove-button" }),
    );

    expect(setCart).toHaveBeenCalled();
    expect(cart).toMatchObject({});

    rerender(
      <CartActions product={mockProduct} cart={cart} setCart={setCart} />,
    );

    expect(
      screen.getByRole("button", { name: "cart-add-button" }),
    ).toBeInTheDocument();
  });

  it("removes item from the cart when remove from cart is clicked", async () => {
    const user = userEvent.setup();

    let cart = { ...mockCart };
    let setCart = vi.spyOn(
      {
        setCart: (newCart) => {
          if (typeof newCart === "function") {
            cart = newCart(cart);
          } else {
            cart = newCart;
          }
        },
      },
      "setCart",
    );

    const { rerender } = render(
      <CartActions product={mockProduct} cart={cart} setCart={setCart} />,
    );

    await user.click(
      screen.getByRole("button", { name: "cart-remove-button" }),
    );

    expect(setCart).toHaveBeenCalled();
    expect(cart).toMatchObject({});

    rerender(
      <CartActions product={mockProduct} cart={cart} setCart={setCart} />,
    );

    expect(
      screen.getByRole("button", { name: "cart-add-button" }),
    ).toBeInTheDocument();
  });

  it("decrements items from the cart", async () => {
    const user = userEvent.setup();

    let cart = { 7: 2 };
    let setCart = vi.spyOn(
      {
        setCart: (newCart) => {
          if (typeof newCart === "function") {
            cart = newCart(cart);
          } else {
            cart = newCart;
          }
        },
      },
      "setCart",
    );

    render(<CartActions product={mockProduct} cart={cart} setCart={setCart} />);

    await user.click(
      screen.getByRole("button", { name: "quantity-decrement-button" }),
    );

    expect(setCart).toHaveBeenCalled();
    expect(cart).toMatchObject({ 7: 1 });
  });

  it("removes items from cart when decremented to 0", async () => {
    const user = userEvent.setup();

    let cart = { 7: 1 };
    let setCart = vi.spyOn(
      {
        setCart: (newCart) => {
          if (typeof newCart === "function") {
            cart = newCart(cart);
          } else {
            cart = newCart;
          }
        },
      },
      "setCart",
    );

    render(<CartActions product={mockProduct} cart={cart} setCart={setCart} />);

    await user.click(
      screen.getByRole("button", { name: "quantity-decrement-button" }),
    );

    expect(setCart).toHaveBeenCalled();
    expect(cart).toMatchObject({});
  });

  it("increments items on the cart", async () => {
    const user = userEvent.setup();

    let cart = { 7: 2 };
    let setCart = vi.spyOn(
      {
        setCart: (newCart) => {
          if (typeof newCart === "function") {
            cart = newCart(cart);
          } else {
            cart = newCart;
          }
        },
      },
      "setCart",
    );

    render(<CartActions product={mockProduct} cart={cart} setCart={setCart} />);

    await user.click(
      screen.getByRole("button", { name: "quantity-increment-button" }),
    );

    expect(setCart).toHaveBeenCalled();
    expect(cart).toMatchObject({ 7: 3 });
  });

  it("changes items count on the cart with custom input", async () => {
    const user = userEvent.setup();

    let cart = { ...mockCart };
    let setCart = vi.spyOn(
      {
        setCart: (newCart) => {
          if (typeof newCart === "function") {
            cart = newCart(cart);
          } else {
            cart = newCart;
          }
        },
      },
      "setCart",
    );

    render(<CartActions product={mockProduct} cart={cart} setCart={setCart} />);

    await user.clear(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    );
    await user.type(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
      "256",
    );

    expect(setCart).toHaveBeenCalled();
    expect(cart).toMatchObject({ 7: 256 });
  });

  it("removes item from cart when custom input is 0", async () => {
    const user = userEvent.setup();

    let cart = { ...mockCart };
    let setCart = vi.spyOn(
      {
        setCart: (newCart) => {
          if (typeof newCart === "function") {
            cart = newCart(cart);
          } else {
            cart = newCart;
          }
        },
      },
      "setCart",
    );

    render(<CartActions product={mockProduct} cart={cart} setCart={setCart} />);

    await user.clear(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
    );
    await user.type(
      screen.getByRole("textbox", { name: "custom-quantity-input" }),
      "0",
    );

    expect(setCart).toHaveBeenCalled();
    expect(cart).toMatchObject({});
  });
});
