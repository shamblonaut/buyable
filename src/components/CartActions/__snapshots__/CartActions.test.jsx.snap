// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CartActions component > renders correctly when cart is empty 1`] = `
<div>
  <div
    class="sc-bjMIFn iUmaXL"
  >
    <button
      aria-label="cart-add-button"
      class="sc-dClGHI sc-fhfEft eEFmxf jwigen"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-shopping-cart"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="8"
          cy="21"
          r="1"
        />
        <circle
          cx="19"
          cy="21"
          r="1"
        />
        <path
          d="M2.05 2.05h2l2.66 12.42a2 2 0 0 0 2 1.58h9.78a2 2 0 0 0 1.95-1.57l1.65-7.43H5.12"
        />
      </svg>
       Add to cart
    </button>
  </div>
</div>
`;

exports[`CartActions component > renders correctly when cart is not empty 1`] = `
<div>
  <div
    class="sc-bjMIFn iUmaXL"
  >
    <div
      aria-label="cart-quantity-selector"
      class="sc-eQwNpu czpjZa"
    >
      <button
        aria-label="quantity-decrement-button"
        class="sc-gvqKNf sc-cOpnSz iwcSHy bQiQpy"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-minus"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12h14"
          />
        </svg>
      </button>
      <input
        aria-label="custom-quantity-input"
        class="sc-eVqvcJ fQmksd"
        type="text"
        value="13"
      />
      <button
        aria-label="quantity-increment-button"
        class="sc-gvqKNf sc-kjwdDK iwcSHy chzemP"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-plus"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12h14"
          />
          <path
            d="M12 5v14"
          />
        </svg>
      </button>
    </div>
    <button
      aria-label="cart-remove-button"
      class="sc-dClGHI sc-bAehkN eEFmxf bdzNYM"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-trash2 lucide-trash-2"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 6h18"
        />
        <path
          d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
        />
        <path
          d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
        />
        <line
          x1="10"
          x2="10"
          y1="11"
          y2="17"
        />
        <line
          x1="14"
          x2="14"
          y1="11"
          y2="17"
        />
      </svg>
    </button>
  </div>
</div>
`;
